### PR TITLE
[MIN-76] Implement a metric database

### PR DESCRIPTION
**Pull request by osw282**
This PR introduces two methods for running the monitoring service: locally or on a Kubernetes cluster.

We achieved this by Dockerizing the metric service through a custom Dockerfile. This enables the entire monitoring service, including the PostgreSQL database and the metric service, to be run locally using docker-compose.

For deployment on a Kubernetes cluster, we've imeplemted three Kubernetes manifests. These are written for deployment of both the monitoring service and the PostgreSQL database. 

Given that Flask servers aren't intended for production use, we are also employing gunicorn as a WSGI server.

We've also updated the README to guide users on running the monitoring service in either of the two environments - locally or remotely on a Kubernetes cluster.

NOTE:
New dependencies added, run `poetry install`.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -193,41 +193,82 @@ If you visit that URL in browser, you should be able to interact with the deploy
...
+curl localhost:5000/query_readability
 
 or
 
-curl http://127.0.0.1:5000//query_embedding_drift
+curl localhost:5000//query_embedding_drift
+```
+
+### Running on k8s cluster
+To run the monitoring service on k8s, `matcha provision` must be run before hand. We will need to build and push the metric service application to ACR. This image will be used by Kubernetes deployment. Before that, we need to set two bash variables, one for ACR registry URI and another for ACR registry name. We will use matcha get command to do this.
```
Comment (JonoCX) [inline README.md]: Before hand -> beforehand

File: README.md [hunk 2]
```diff
@@ -193,41 +193,82 @@ If you visit that URL in browser, you should be able to interact with the deploy
...
+docker push $acr_registry_uri/monitoring:latest
+```
+
+Next, we apply the Kubernetes manifest to deploy the metric service and the metric database on AKS.
+
+```bash
+kubectl apply -f infrastructure/monitoring
+```
+
+Finally, once the pod is running, we verify that our monitoring service is working. The command below should provide an IP address for the metric service interface.
```
Comment (JonoCX) [inline README.md:202]: Spacing?

File: README.md [hunk 3]
```diff
@@ -193,41 +193,82 @@ If you visit that URL in browser, you should be able to interact with the deploy
...
+```bash
+kubectl get pods # Checking whether the monitoring pod is running
+
+kubectl get svc monitoring-service -o jsonpath='{.status.loadBalancer.ingress[0].ip}
```
Comment (JonoCX) [inline README.md]: The quote isn't closed

File: README.md [hunk 4]
```diff
@@ -193,41 +193,82 @@ If you visit that URL in browser, you should be able to interact with the deploy
...
+```
+
+We should be able to curl the external IP returned running the above command at port 5000.
+```bash
+curl {external-ip:5000}
+
+# The response should be:
+Hello world from the metric service.%
```
Comment (JonoCX) [inline README.md]: %?

**Review by JonoCX — APPROVED** [review]
LGTM! Nice work Oscar 🦖

File: README.md (hunk 2 above)
Comment (osw282) [reply README.md:202]: Like removing the space?

File: README.md [hunk 5]
```diff
@@ -193,41 +193,82 @@ If you visit that URL in browser, you should be able to interact with the deploy
...
-export DB_NAME=<The database name>
-export DB_HOST=<The database host>
-export DB_USER=<The database user>
-export DB_PASSWORD=<The database password, export DB_PASSWORD=None otherwise.>
-export DB_PORT=<The database port>
+ENV DB_HOST=localhost
+
+# should be changed to
+
+ENV DB_HOST=metric-db
```
Comment (d-lowl) [inline README.md]: Why can't this be set by default?

File: monitoring/metric_service/Dockerfile [hunk 6]
```diff
@@ -0,0 +1,28 @@
+FROM python:3.10-slim-buster
+
+WORKDIR /app
+
+ENV VIRTUAL_ENV=/opt/venv
+RUN python3 -m venv $VIRTUAL_ENV
+ENV PATH="$VIRTUAL_ENV/bin:$PATH"
+
+ENV DB_NAME=postgres
+# For the below ENV DB_HOST, change value to "metric-db" if running locally using docker-compose
```
Comment (d-lowl) [inline monitoring/metric_service/Dockerfile]: Can we not set this in the docker compose itself?

**Review by d-lowl — CHANGES_REQUESTED** [review]
I suppose we can simplify the instructions a bit, by moving configuration to docker-compose itself. Other than that, looks good

File: README.md (hunk 5 above)
Comment (osw282) [reply README.md]: Running the metric service and the database locally with docker compose uses two containers, I don't think metric service container can reach the database container using localhost. I might be wrong tho.

Comment (d-lowl) [reply README.md]: Well, setting it in docker-compose does the trick

**Review by d-lowl — APPROVED** [review]

**Pull request merged**